server/src/main/java/org/elasticsearch/indices/IndexingMemoryController.java

```diff
@@ -216,11 +242,18 @@ protected void deactivateThrottling(IndexShard shard) {
     @Override
     public void postIndex(ShardId shardId, Engine.Index index, Engine.IndexResult result) {
         recordOperationBytes(index, result);
+        // Piggy back on indexing threads to write segments. We're not submitting a task to the index threadpool because we want memory to
+        // be reclaimed rapidly. This has the downside of increasing the latency of _bulk requests though. Lucene does the same thing in
+        // DocumentsWriter#postUpdate, flushing a segment because the size limit on the RAM buffer was reached happens on the call to
+        // IndexWriter#addDocument.
+        writePendingIndexingBuffers();
```

Comment:
Here we could be penalizing an unrelated write operation with flushing multiple large segments. Lucene's write stays within the shard, here we go cross-shard/index.

I am not sure it will cause large disruptions though. But I wonder if it was worth only flushing the specific shard that the original write concerns?

Reply:
I wondered about that but it fels like there could be adversarial cases that I didn't feel good about. E.g. say you have two shards S1 and S2. S1 uses 90% of the indexing buffer while S2 uses 9%. But suddenly, S1 stops indexing and only S2 indexes. You might never be able to flush the indexing buffer of S1 if you forced flushes to only happen on an indexing thread for the same index?